liblepton/scheme/conf/upgrade.scm

```diff
@@ -0,0 +1,233 @@
+      ( set! keys (conf-keys-list) )
+      ( set! data
+        ( read-old cfg keys #:report-absent-keys report-absent-keys )
+      )
+    )
+    ( lambda( ex . args )
+      ( throw 'load "Cannot read config:" ex args )
+    )
+  )
+
+
+  ;
+  ; write new config file (lepton*.conf):
+  ;
+
+  ( format #t "~%" )
+
+  ( config-set-legacy-mode! #f ) ; use lepton*.conf files
+
+  ( set! cfg ( get-cfg-ctx what ) )
+  ( unless cfg
+    ( throw 'ctx "Cannot get config context (output)" )
+  )
+
+  ( set! fname (config-filename cfg) )
+  ( when ( and fname (access? fname F_OK) (not overwrite) )
+    ( throw 'outfile "Output config file already exists:" fname )
+  )
+
+  ( format #t "ii: OUTPUT: ~a~%" fname )
+
+  ( catch #t
+    ( lambda()
+      ( write-new cfg data )
+      ( config-save! cfg )
+    )
+    ( lambda( ex . args )
+      ( throw 'save "Cannot write config:" ex args )
+    )
+  )
+
+  ; return:
+  fname
+
+) ; let
+) ; convert()
+
+
+
+; get configuration context
+;
+; [what]: symbol: 'local, 'user or 'system
+; {ret}:  requested configuration context or #f on failure
+;
+( define ( get-cfg-ctx what )
+  ; return:
+  ( cond
+    (  ( eq? what 'local  )  ( path-config-context (getcwd) )  )
+    (  ( eq? what 'user   )  ( user-config-context )           )
+    (  ( eq? what 'system )  ( system-config-context )         )
+    (  else #f  )
+  )
+)
+
+
+
+; read legacy configuration
+;
+; [cfg]:                 configuration context to read from
+; [keys]:                cfg keys list as defined in (conf keys) module
+; [report-absent-keys]:  report absent keys
+; {ret}:                 list of entries: ( grp-new key-new val )
+;
+( define* ( read-old cfg keys #:key (report-absent-keys #f) )
+( let
+  (
+  ( grp-old  #f )
+  ( key-old  #f )
+  ( grp-new  #f )
+  ( key-new  #f )
+  ( pfn      #f )
+  ( val      #f )
+  ( res     '() )
+  )
+
+  ( define ( parse-entry e )
+    ( set! pfn     (list-ref e 0) )
+    ( set! grp-old (list-ref e 2) )
+    ( set! key-old (list-ref e 3) )
+    ( set! grp-new (list-ref e 4) )
+    ( set! key-new (list-ref e 5) )
+  )
+
+  ( define ( add-to-res g k v ) ; group, key, value
+    ( set! res ( cons (list g k v) res ) )
+  )
+
+  ( for-each
+  ( lambda( entry )
+
+    ( parse-entry entry )
+
+    ( catch #t
+      ( lambda()
+        ( set! val ( pfn cfg grp-old key-old ) ) ; pfn() throws
+        ( format #t "ii: read   [~a]::~a ~65,4t = [~a]~%" grp-old key-old val )
+        ( add-to-res grp-new key-new val )
+      )
+      ( lambda( ex . args )
+        ( if report-absent-keys
+          ( format #t "ww: !read  [~a]::~a~%" grp-old key-old )
+        )
+      )
+    )
+
+  )
+  keys
+  )
+
+  ; return:
+  res
+
+) ; let
+) ; read-old()
+
+
+
+; write new configuration
+;
+; [cfg]: configuration context to write to
+; [res]: list returned by read-old()
+;
+( define ( write-new cfg res )
+( let
+  (
+  ( grp-new #f )
+  ( key-new #f )
+  ( val     #f )
+  )
+
+  ( define ( parse-entry e )
```

Comment:
Please consider using of `match` from the `(ice-9 match)` module for such parse functions. There are examples in `lepton-netlist` code.

Reply:
Didn't know about `match()`. :-) It is really helpful function, thank you.